### PR TITLE
SONAR-6336 Fail to update or create a condition that already exists

### PR DESCRIPTION
**Pull request by julienlancelot**
SONAR-6336 Fail to update or create a condition that already exists

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java [hunk 1]
```diff
@@ -71,7 +76,7 @@
   private final Settings settings;
 
   public QualityGates(QualityGateDao dao, QualityGateConditionDao conditionDao, MetricFinder metricFinder, PropertiesDao propertiesDao, ComponentDao componentDao,
-                      MyBatis myBatis, UserSession userSession, Settings settings) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: very minor, but I wonder what is the best indentation. I have the same configuration, but I find the old one better

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java [hunk 2]
```diff
@@ -263,17 +267,31 @@ public boolean currentUserHasWritePermission() {
...
     }
   }
 
+  private void checkConditionDoesNotAlreadyExistOnSameMetricAndPeriod(long qGateId, Metric metric, @Nullable final Integer period, Errors errors) {
+    Collection<QualityGateConditionDto> conditions = conditionDao.selectForQualityGate(qGateId);
+    if (conditions.isEmpty()) {
+      return;
+    }
+    boolean conditionExists = from(conditions).anyMatch(new MatchQualityGate(metric.getId(), period));
+    if (period == null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: I would use the ternary operator to define the error message and then have only one error check. Something similar to:
```
String errorMessage = period==null
  ? format("Condition on metric '%s' already exists.", metric.getName())
  : format("Condition on metric '%s' over leak period already exists.", metric.getName());
errors.check(!conditionExists, errorMessage);
```

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java [hunk 3]
```diff
@@ -377,4 +394,22 @@ private void checkPermission(Long projectId, DbSession session) {
       throw new ForbiddenException("Insufficient privileges");
     }
   }
+
+  private static class MatchQualityGate implements Predicate<QualityGateConditionDto> {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: rename `MatchMetricAndPeriod`

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: As long as we're all using the same one, it's good to me :)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java (hunk 2 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java (hunk 3 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: Right !

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java [hunk 4]
```diff
@@ -449,8 +458,39 @@ public void should_fail_create_condition_on_invalid_period() {
...
+    addMetric(metricKey, "Coverage");
+    when(dao.selectById(QUALITY_GATE_ID)).thenReturn(new QualityGateDto().setId(QUALITY_GATE_ID));
+    when(conditionDao.selectForQualityGate(QUALITY_GATE_ID)).thenReturn(
+      singletonList(new QualityGateConditionDto().setMetricKey(metricKey).setMetricId(METRIC_ID)));
+
+    expectedException.expect(BadRequestException.class);
+    expectedException.expectMessage("Condition on metric 'Coverage' already exists.");
+    qGates.createCondition(QUALITY_GATE_ID, metricKey, "LT", "90", null, null);
+
+    verify(conditionDao, never()).insert(any(QualityGateConditionDto.class));
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java]: it fails, this line shouldn't be reachable

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java [hunk 5]
```diff
@@ -449,8 +458,39 @@ public void should_fail_create_condition_on_invalid_period() {
...
+    when(dao.selectById(QUALITY_GATE_ID)).thenReturn(new QualityGateDto().setId(QUALITY_GATE_ID));
+
+    when(conditionDao.selectForQualityGate(QUALITY_GATE_ID)).thenReturn(
+      singletonList(new QualityGateConditionDto().setMetricKey(metricKey).setMetricId(METRIC_ID).setPeriod(1)));
+
+    expectedException.expect(BadRequestException.class);
+    expectedException.expectMessage("Condition on metric 'New Coverage' over leak period already exists.");
+    qGates.createCondition(QUALITY_GATE_ID, metricKey, "LT", "90", null, 1);
+
+    verify(conditionDao, never()).insert(any(QualityGateConditionDto.class));
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java]: it fails, this line shouldn't be reachable

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java]: indeed, I'll remove it

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java [hunk 6]
```diff
@@ -377,4 +394,22 @@ private void checkPermission(Long projectId, DbSession session) {
...
+    @CheckForNull
+    private final Integer period;
+
+    public MatchQualityGate(int metricId, @Nullable Integer period) {
+      this.metricId = metricId;
+      this.period = period;
+    }
+
+    @Override
+    public boolean apply(@Nonnull QualityGateConditionDto input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java [hunk 7]
```diff
@@ -449,8 +458,39 @@ public void should_fail_create_condition_on_invalid_period() {
...
+  public void fail_to_create_condition_when_condition_on_same_metric_already_exist() throws Exception {
+    String metricKey = "coverage";
+    addMetric(metricKey, "Coverage");
+    when(dao.selectById(QUALITY_GATE_ID)).thenReturn(new QualityGateDto().setId(QUALITY_GATE_ID));
+    when(conditionDao.selectForQualityGate(QUALITY_GATE_ID)).thenReturn(
+      singletonList(new QualityGateConditionDto().setMetricKey(metricKey).setMetricId(METRIC_ID)));
+
+    expectedException.expect(BadRequestException.class);
+    expectedException.expectMessage("Condition on metric 'Coverage' already exists.");
+    qGates.createCondition(QUALITY_GATE_ID, metricKey, "LT", "90", null, null);
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java]: rename `qGates` to `underTest`

Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/QualityGatesTest.java]: ok

**Comment by teryk** [comment]
apply feedbacks and LGTM!

**Pull request closed without merging**